### PR TITLE
Separate PageContext and ViewContext

### PR DESCRIPTION
**Pull request by rynowak**
This change decouples PageContext and ViewContext completely.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/IPageActivatorProvider.cs [hunk 1]
```diff
@@ -15,13 +16,13 @@ public interface IPageActivatorProvider
         /// </summary>
         /// <param name="descriptor">The <see cref="CompiledPageActionDescriptor"/>.</param>
         /// <returns>The delegate used to activate the page.</returns>
-        Func<PageContext, object> CreateActivator(CompiledPageActionDescriptor descriptor);
+        Func<PageContext, ViewContext, object> CreateActivator(CompiledPageActionDescriptor descriptor);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/IPageActivatorProvider.cs]: Basically, the `Page` gets the `PageContext` and `ViewContext` - the `PageModel` just gets the `PageContext`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 2]
```diff
@@ -360,44 +424,17 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
...
-                        executor = CacheEntry.Executors[i];
-                        break;
-                    }
-                }
-
-                var instance = actionDescriptor.ModelTypeInfo == actionDescriptor.HandlerTypeInfo ? _model : _page;
-                result = await executor(instance, arguments);
+                // If we get here we're going to render the page so we need to initialize the context.
+                pageResult.Page = _page;
+                pageResult.ViewData = pageResult.ViewData ?? _pageContext.ViewData;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: This is the one part I thought was gross, we need to special case `PageResult `here. The `pageResult.ViewData` will go away in my next round a of changes.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 3]
```diff
@@ -432,6 +469,53 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
...
+            }
+
+            var viewStarts = new IRazorPage[CacheEntry.ViewStartFactories.Count];
+            for (var i = 0; i < viewStarts.Length; i++)
+            {
+                var pageFactory = CacheEntry.ViewStartFactories[i];
+                viewStarts[i] = pageFactory();
+            }
+
+            return viewStarts;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: I'm going to move this to the result executor.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/DefaultPageActivator.cs [hunk 4]
```diff
@@ -12,10 +13,10 @@ namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
     /// </summary>
     public class DefaultPageActivator : IPageActivatorProvider
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/DefaultPageActivator.cs]: If you feel up to it, can you change the type name to `DefaultPageActivatorProvider`?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 5]
```diff
@@ -303,47 +321,99 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
...
             {
-                _model = CacheEntry.ModelFactory(_pageContext);
+                // If we get here, we are going to render the page, so we need to create it and then initialize
+                // the context so we can run the result.
+                _viewContext = new ViewContext(
+                    _pageContext,
+                    NullView.Instance,
+                    _pageContext.ViewData,
+                    _tempDataFactory.GetTempData(_pageContext.HttpContext),
+                    TextWriter.Null,
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: Should we make a ctor without the `TextWriter`? I don't think we initialize it to the right value either here or in `VIewExecutor`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 6]
```diff
@@ -303,47 +321,99 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
...
+                _pageContext,
+                NullView.Instance,
+                _pageContext.ViewData,
+                _tempDataFactory.GetTempData(_pageContext.HttpContext),
+                TextWriter.Null,
+                _htmlHelperOptions);
+
+            _page = (Page)CacheEntry.PageFactory(_pageContext, _viewContext);
+
+            if (_actionDescriptor.ModelTypeInfo == _actionDescriptor.PageTypeInfo)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: Shouldn't need this check

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PassThruRazorPageActivator.cs [hunk 7]
```diff
@@ -22,10 +22,10 @@ public void Activate(IRazorPage page, ViewContext context)
             var razorView = (RazorView)context.View;
             if (ReferenceEquals(page, razorView.RazorPage))
             {
-                var pageContext = (PageContext)context;
+                var actionDescriptor = (CompiledPageActionDescriptor)context.ActionDescriptor;
                 var vddType = typeof(ViewDataDictionary<>);
 
-                var modelTypeInfo = pageContext.ActionDescriptor.ModelTypeInfo ?? pageContext.ActionDescriptor.PageTypeInfo;
+                var modelTypeInfo = actionDescriptor.ModelTypeInfo ?? actionDescriptor.PageTypeInfo;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PassThruRazorPageActivator.cs]: Extra line

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageResult.cs [hunk 8]
```diff
@@ -56,14 +42,14 @@ public PageResult(PageBase page, object model)
         /// <inheritdoc />
         public override Task ExecuteResultAsync(ActionContext context)
         {
-            if (!object.ReferenceEquals(context, Page.PageContext))
+            if (!(context is PageContext pageContext))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageResult.cs]: Is the exception message good enough for people trying to return this result from a controller?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ViewDataDictionaryFactory.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        public static Func<IModelMetadataProvider, ModelStateDictionary, ViewDataDictionary> CreateFactory(TypeInfo modelType)
+        {
+            if (modelType == null)
+            {
+                throw new ArgumentNullException(nameof(modelType));
+            }
+
+            var type = typeof(ViewDataDictionary<>).MakeGenericType(modelType);
+            var constructor = type.GetConstructor(new Type[] { typeof(IModelMetadataProvider), typeof(ModelStateDictionary) });
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ViewDataDictionaryFactory.cs]: Would it be less code if we make a `VDDFactory<T>` (similar to `CollectionModelBinder`)?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ViewDataDictionaryFactory.cs]: No. No one ever has a `T` they have `Type`. This is already super simple, I don't want to spend more time on it.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/DefaultPageActivator.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/DefaultPageActivator.cs]: Sure

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: I don't want to add more ctors to `ViewContext` - there's lots of history here and we spent a lot of time winnowing it down to the current set.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageResult.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageResult.cs]: Yeah good point. I'll mention that explicitly.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: yes I do. The page can use `@model` and have handler methods

**Pull request closed without merging**